### ISSUE TITLE
Performance bug: runs 3x faster if a call to blake3_hasher_update is broken into two calls

### ISSUE THREAD

**Issue by willbryant** [BLAKE3-team/BLAKE3#69]
I've extracted a minimal test case reproducing a very strange performance cliff in my application - strange because it can be worked around in a way that suggests that the buffering algorithm in blake3_hasher_update isn't working properly.

I'm using the C port in my app, I haven't had a chance to try and get the Rust version going and see if it has the same issue, but from reading the two the buffering algorithm looks the same to me.

To produce the problem, we need to start out by hashing something small, eg. a size_t integer (in my real application this would be the length of the following string).  Then follow with a big buffer.

It's not surprising that this would be measurably slower than just hashing the big buffer without the integer first, since it requires retaining a partial block of data.  What is strange is that the _entire operation_ on the big buffer is much slower, not just the hashing of the first full block of data.

And even stranger, you can restore back essentially full performance by breaking the blake3_hasher_update call for the big block of data into two, on the boundary of 4096 bytes minus however many bytes you've already hashed.  (Yes, 4096 - not BLAKE3_CHUNK_LEN, 1024, like any sensible bug would have chosen.)

In fact making _more_ calls to blake3_hasher_update makes the whole thing 3x faster even though the same data is processed in the end.  Obviously, if this is really expected, we should be doing that inside blake3_hasher_update.

Code to reproduce:

```c
#include "blake3.h"
#include <stdlib.h>
#include <strings.h>
#include <stdio.h>
#include <time.h>

int main() {
  blake3_hasher hasher;
  uint8_t output[BLAKE3_OUT_LEN];
  size_t reps = 1000; // arbitrary

  size_t small_value = 0;
  size_t buf_size = 1024*1024; // 1mb; fairly arbitrary, 256kb gives the same result, as does 1024*1024 - sizeof(size_t)
  unsigned char *buf = malloc(buf_size);
  bzero(buf, buf_size);

  // variant 1: hash a size_t, then hash the remainder of the 1mb buffer
  clock_t start1 = clock();
  for (int rep = 0; rep < reps; rep++) {
    blake3_hasher_init(&hasher);
    blake3_hasher_update(&hasher, &small_value, sizeof(small_value));
    blake3_hasher_update(&hasher, buf, buf_size);
    blake3_hasher_finalize(&hasher, output, BLAKE3_OUT_LEN);
  }
  clock_t end1 = clock();
  printf("variant 1 ran in %f, hash ", ((float)(end1 - start1))/CLOCKS_PER_SEC);
  for (size_t i = 0; i < BLAKE3_OUT_LEN; i++) {
    printf("%02x", output[i]);
  }
  printf("\n");

  // variant 2: hash a size_t, then hash 4k minus the size of a size_t, then hash the remainder of the 1mb buffer
  clock_t start2 = clock();
  for (int rep = 0; rep < reps; rep++) {
    blake3_hasher_init(&hasher);
    blake3_hasher_update(&hasher, &small_value, sizeof(small_value));
    size_t magic_amount_to_update = 4096 - sizeof(small_value); // <-- 4096 is the magic number; 1024 (BLAKE3_CHUNK_LEN) does not produce a significant speedup; 2048 produces a partial speedup
    blake3_hasher_update(&hasher, buf, magic_amount_to_update); // <-- new line
    blake3_hasher_update(&hasher, buf + magic_amount_to_update, buf_size - magic_amount_to_update); // <-- remainder of the full buffer
    blake3_hasher_finalize(&hasher, output, BLAKE3_OUT_LEN);
  }
  clock_t end2 = clock();
  printf("variant 2 ran in %f, hash ", ((float)(end2 - start2))/CLOCKS_PER_SEC);
  for (size_t i = 0; i < BLAKE3_OUT_LEN; i++) {
    printf("%02x", output[i]);
  }
  printf("\n");

  return 0;
}
```

On my Macbook Pro I get:

variant 1 ran in 0.946282, hash 80b378d1f0bc7ee2e9c96dc30108187a03779c8dfd951047bf6c99ce88be2ff6
variant 2 ran in 0.302671, hash 80b378d1f0bc7ee2e9c96dc30108187a03779c8dfd951047bf6c99ce88be2ff6

**Comment by oconnor663** [comment]
I'm very glad you caught this, and thank you for the extremely thorough repro. The problem is caused by the optimization described in [this comment](https://github.com/BLAKE3-team/BLAKE3/blob/74b5fe90545678429dcb424983fcab69dc4d9e07/c/blake3.c#L489-L492):

    // Shrink the subtree_len until *half of it* it evenly divides the count so
    // far. Why half? Because compress_subtree_to_parent_node will return a
    // pair of chaining values, each representing half of the input. As long as
    // those evenly divide the input so far, we're good.

That comment is correct as far as it goes. If we've hashed say 1 chunk so far, and then you call update with 2 chunks, processing that whole 2-chunk input with a single call to `compress_subtree_to_parent_node` gives the correct result (and the performance benefits of 2-way SIMD), because that function is going to return the two separate chaining values that comprise a "parent node", and each of those two chaining values gets merged into the tree state separately. However, if that function returned a single chaining value instead, that would lead to an incorrect result, because it's supposed to be chunks `#0` and `#1` that are siblings in the tree, not chunks `#1` and `#2`. Likewise, calling it with 4 chunks in this situation would be incorrect.

What I didn't realize at the time, and what you've discovered, is that that "optimization" gets the update loop stuck in a bad state. If we've hashed 1 chunk so far, and then we get a large update, the calls to `compress_subtree_to_parent_node` will be 2 chunks, 2 chunks, 2 chunks, ... forever. The total number of chunks so far will remain odd, and the "half of it evenly divides the count so far" rule will always give the answer 2. What we _should_ be doing instead is 1, 2, 4, 8, ..., so that the total number of chunks becomes divisible by larger powers of 2, and we can compress larger subtrees with higher degrees of parallelism.

The fix is going to be simple. We should just stop halving the subtree length in that part of the update loop:

```diff
diff --git a/c/blake3.c b/c/blake3.c
index b4a5992..d28b7f3 100644
--- a/c/blake3.c
+++ b/c/blake3.c
@@ -495,7 +495,7 @@ void blake3_hasher_update(blake3_hasher *self, const void *input,
     // consistently passes power-of-2 inputs of the same size, as is hopefully
     // typical, this loop condition will always fail, and subtree_len will
     // always be the full length of the input.)
-    while ((((uint64_t)((subtree_len / 2) - 1)) & count_so_far) != 0) {
+    while ((((uint64_t)(subtree_len - 1)) & count_so_far) != 0) {
       subtree_len /= 2;
     }
     // The shrunken subtree_len might now be 1 chunk long. If so, hash that one
```

I haven't measured the Rust code yet, but it uses the same strategy, and I'm pretty sure it has the same problem. I'll have the fix for both out shortly. Again, so glad you caught this.
